third_party/move/move-vm/runtime/src/storage/publishing.rs

```diff
@@ -210,32 +229,74 @@ impl<'a, M: ModuleStorage> StagingModuleStorage<'a, M> {
         };
 
         // Finally, verify the bundle, performing linking checks for all staged modules.
-        for (addr, name) in staged_module_storage
+        let staged_runtime_environment = staged_module_storage.runtime_environment();
+        for (addr, name, bytes, compiled_module) in staged_module_storage
             .storage
             .byte_storage()
-            .staged_module_bytes
+            .staged_modules
             .iter()
             .flat_map(|(addr, account_storage)| {
                 account_storage
                     .iter()
-                    .map(move |(name, _)| (addr, name.as_ident_str()))
+                    .map(move |(name, (bytes, module))| (addr, name, bytes, module))
             })
         {
-            // Verify the module and its dependencies, and that they do not form a cycle.
-            let module = staged_module_storage
-                .fetch_verified_module(addr, name)?
-                .ok_or_else(|| {
-                    PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
-                        .with_message(format!(
-                            "Staged module {}::{} must always exist",
-                            addr, name
-                        ))
-                        .finish(Location::Undefined)
-                })?;
+            if is_lazy_loading_enabled {
+                // Local bytecode verification.
+                staged_runtime_environment.paranoid_check_module_address_and_name(
+                    compiled_module,
+                    compiled_module.self_addr(),
+                    compiled_module.self_name(),
+                )?;
+                let locally_verified_code = staged_runtime_environment
+                    .build_locally_verified_module(
+                        compiled_module.clone(),
+                        bytes.len(),
+                        &sha3_256(bytes),
+                    )?;
+
+                // Linking checks to immediate dependencies. Note that we do not check cyclic
+                // dependencies here.
+                let mut verified_dependencies = vec![];
+                for (dep_addr, dep_name) in locally_verified_code.immediate_dependencies_iter() {
+                    // MODULE METERING SAFETY:
+                    //   Immediate dependency of the module in a bundle must be metered at the
+                    //   caller side.
+                    traversal_context
+                        .check_is_special_or_visited(dep_addr, dep_name)
+                        .map_err(|err| err.finish(Location::Undefined))?;
+                    let dependency =
+                        staged_module_storage.fetch_existing_verified_module(dep_addr, dep_name)?;
```

Comment:
Note: ignore this call: next PR will ensure this call does not load the transitive closure but resolves dependencies lazily.